### ISSUE TITLE
Introduce a character limit on email templates

### ISSUE THREAD

**Issue by yaelberger-commits** [cds-snc/notification-planning#1233]
## Introduce a size limit on email templates

As a Notify Sender, I need to be able to send long emails successfully so that I don't have failed, delayed or duplicate emails.

As Notify admin, I need to be able to limit the size of an email template so that a template doesn't cause failures, delays or duplicate sends in our system when the db times out when trying to save notifications. 

As a recipient, I need emails from the GoC to be readable so that I don't break my brain trying to read them

https://docs.google.com/document/d/1RuHawkzLELYTSR1RqTRWBrR1t6xneIXlarMjLtc0R7o/edit#

WHY are we building?
We've had a couple incidents where very long templates cause problems for our system. 

WHAT are we building?
An email template size (character - since the issue comes from there being a lot of plain text for the database to process) limit - performance testing to understand if something like 50k would significantly reduce the failure rate could be done later on.
Associated content/design (error message and rest to be designed through story mapping activity)
For this card, the scope excludes designing the 'right' character limit for Recipients.

VALUE created by our solution
We won't have failed, delayed or duplicate emails because they time out when saving to the database
Shorter emails for Recipients

## Documentation and Artifacts

Good docs, figma mockups, ADRs, screenshots etc.

The template that caused Notify's performance issues can be found [here](https://staging.notification.cdssandbox.xyz/services/1dac916a-05b2-49ca-bf75-072231c3f505/templates/04a71bf5-f302-40d7-86d9-d357f05cfc02).

The template is about 227 KB, or 210,166 characters long. We should enforce a limit that is less than this - maybe 50k characters?

Here is the error message we are using to enforce an SMS template limit:
<img src="https://images.zenhubusercontent.com/63176b6f863fa9b22e9eee20/c91d64a1-2cd9-48d8-89c9-a82f42a43f75" width="500">

## Acceptance Criteria
It is not possible to send an email template longer than X.

_For this card, the scope excludes designing the ‘right’ character limit for Recipients._

- [x] An email template size (character - since the issue comes from there being a lot of plain text for the database to process) limit
- [x] Performance testing to understand if something like 50k would significantly reduce the failure rate.
- [x] Associated content/design (error message and rest to be designed through story mapping activity)
- [x] Generate appropriate log messages so that executions of this feature can be tracked
- [x] A character limit is applied to all email templates (New only)

## Related [Research Airtable](https://airtable.com/appWwAmHwDLtpIyko/tbl38n7ofWYBuezFc/viwQUX3tUxbMxYS7C?blocks=hide) records
- [Users worried about the speed and limit of emails they could send](https://airtable.com/appWwAmHwDLtpIyko/tbl38n7ofWYBuezFc/viwBYyaQxOttZ0IAw/reccFNmVkmM8ykzSP?blocks=hide)
- [Participants found the review message page reassuring and useful. A couple participants were worried about the length of the page (especially if they had a long list of recipients)](https://airtable.com/appWwAmHwDLtpIyko/tbl38n7ofWYBuezFc/viwBYyaQxOttZ0IAw/recS7D4BLxWOYvDLJ?blocks=hide)

## QA Steps
- [ ] Tested in a realistic production scenario - a bulk send of 10k emails using a template that is just inside the new limit should not cause errors or performance issues on Staging 
- [ ] Verify that email templates over 50k characters cannot be created

**Comment by yaelberger-commits** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/Notify-Planning-614b3ad91bc2030015ed22f5/issues/cds-snc/notification-planning/1233?planning-poker) with Zenhub @andrewleith @jzbahrai @smcmurtry @whabanks @YedidaZalik @amazingphilippe

**Comment by smcmurtry** [comment]
@YedidaZalik The email template that caused the issues was about 210 thousand characters long, or about 116 pages of content if a standard page has around 1800 characters.

I think this limit should be something like 50 thousand characters, or 27 pages. So almost no one will actually reach this limit.

**Comment by yaelberger-commits** [comment]
@smcmurtry sync with designers on error messages in EN/FR

**Comment by andrewleith** [comment]
- Progress is going well - one PR will be ready for review today; @whabanks to review

**Comment by andrewleith** [comment]
- Work is continuing; validation code is merged
- Still working on validation messages and then they will go to translation
- This work affects Forms so we will need to be careful what we introduce here does not affect them

**Comment by smcmurtry** [comment]
The new error message and translation are visible on staging.
I ran a test with a bulk send of 8k emails that were 50k characters each on staging. No alarms occurred in #notification-staging-ops.

**Comment by smcmurtry** [comment]
One last PR to add logging when someone hits the new email template character limit.

**Comment by smcmurtry** [comment]
This is the state of the email length limits:
- (new) limit to prevent the creation or update of email templates over 50,000 characters
- (existing) limit to prevent sending a one-off email with a variable over 51200 bytes (in this case bytes seems to equal characters)
- (existing) limit to prevent variables in a bulk send over 131072 (bytes/characters)

I also tried an experiment to include multiple variables in a template to see if I could get up to a message length of 200k characters. Then I requested a bulk send of 6000 emails like that. But my computer ran out of memory while making the request, so it didn’t go through.
So the conclusion I’m drawing here is that the new limit plus the existing limits add up to a much lower likelyhood we will have a repeat of this incident.
